trin-storage/src/config.rs

```diff
@@ -0,0 +1,81 @@
+use std::path::PathBuf;
+
+use discv5::enr::NodeId;
+use ethportal_api::types::cli::{BEACON_NETWORK, HISTORY_NETWORK, STATE_NETWORK};
+use r2d2::Pool;
+use r2d2_sqlite::SqliteConnectionManager;
+
+use crate::{error::ContentStoreError, utils::setup_sql, DistanceFunction};
+
+const BYTES_IN_MB_U64: u64 = 1000 * 1000;
+
+/// Factory for creating [PortalStorageConfig] instances
+pub struct PortalStorageConfigFactory {
+    node_id: NodeId,
+    node_data_dir: PathBuf,
+    total_capacity_mb: u64,
+    total_capacity_weight: u64,
+    sql_connection_pool: Pool<SqliteConnectionManager>,
+}
+
+impl PortalStorageConfigFactory {
+    const HISTORY_CAPACITY_WEIGHT: u64 = 1;
+    const STATE_CAPACITY_WEIGHT: u64 = 99;
+    const BEACON_CAPACITY_WEIGHT: u64 = 0; // Beacon doesn't care about given capacity
+
+    pub fn new(
+        total_capacity_mb: u64,
+        subnetworks: &[String],
+        node_id: NodeId,
+        node_data_dir: PathBuf,
+    ) -> Result<Self, ContentStoreError> {
+        let total_capacity_weight = subnetworks
+            .iter()
+            .map(|subnetwork| Self::get_capacity_weight(subnetwork))
+            .sum();
+
+        let sql_connection_pool = setup_sql(&node_data_dir)?;
+
+        Ok(Self {
+            node_data_dir,
+            node_id,
+            total_capacity_mb,
+            total_capacity_weight,
+            sql_connection_pool,
+        })
+    }
+
+    pub fn create(&self, subnetwork: &str) -> PortalStorageConfig {
+        let capacity_weight = Self::get_capacity_weight(subnetwork);
+        let capacity_bytes = if self.total_capacity_weight == 0 {
+            0
+        } else {
+            BYTES_IN_MB_U64 * self.total_capacity_mb * capacity_weight / self.total_capacity_weight
+        };
+        PortalStorageConfig {
+            storage_capacity_bytes: capacity_bytes,
+            node_id: self.node_id,
+            node_data_dir: self.node_data_dir.clone(),
+            distance_fn: DistanceFunction::Xor,
+            sql_connection_pool: self.sql_connection_pool.clone(),
+        }
+    }
+
+    fn get_capacity_weight(subnetwork: &str) -> u64 {
+        match subnetwork {
+            HISTORY_NETWORK => Self::HISTORY_CAPACITY_WEIGHT,
+            STATE_NETWORK => Self::STATE_CAPACITY_WEIGHT,
+            BEACON_NETWORK => Self::BEACON_CAPACITY_WEIGHT,
+            _ => panic!("Invalid subnetwork: {subnetwork}"),
```

Comment:
Is it worth creating a `Subnetwork` enum type? Something like [this](https://github.com/ethereum/trin/blob/master/ethportal-api/src/types/cli.rs#L32) that will cover [these](https://github.com/ethereum/trin/blob/master/ethportal-api/src/types/cli.rs#L19-L21)? So we can skip all the string ser/de?

Reply:
And if so, maybe also for `Network` (mainnet/angelfood)... working with enums is a bit nicer than using raw strings in various places

Reply:
I was thinking the same, but I want to get this pr asap (so I can get another pr before the release this week). So I will leave this for the future (created #1392 ).